### PR TITLE
Improve workers API, fix lost sync.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: runtime/src/main/kotlin/konan/worker/Worker.kt [hunk 1]
```diff
@@ -100,18 +100,13 @@ class Future<T> internal constructor(val id: FutureId) {
...
 
-    override fun equals(other: Any?): Boolean {
-        return (other is Future<*>) && (id == other.id)
-    }
+    override fun equals(other: Any?) = (other is Future<*>) && (id == other.id)
 
-    override fun hashCode(): Int {
-        return id
-    }
+    override fun hashCode() = id as Int
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/Worker.kt]: What's the reason of adding `as Int`?

File: runtime/src/main/kotlin/konan/worker/Worker.kt [hunk 2]
```diff
@@ -240,4 +235,12 @@ internal fun ThrowWorkerInvalidState(): Unit =
         throw IllegalStateException("Illegal transfer state")
 
 @ExportForCppRuntime
-internal fun WorkerLaunchpad(function: () -> Any?) = function()
+internal fun WorkerLaunchpad(function: () -> Any?): Any? {
+    try {
+        return function()
+    } catch (e: Throwable) {
+        println("worker thrown an exception: $e")
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/kotlin/konan/worker/Worker.kt]: Has the worker thrown this exception, or it was the thread sending the job to the worker?

File: runtime/src/main/kotlin/konan/worker/Worker.kt (hunk 1 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/Worker.kt]: Nothing special, just code readability.

Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/Worker.kt]: Removed.

File: runtime/src/main/kotlin/konan/worker/Worker.kt (hunk 2 above)
Comment (olonho) [reply runtime/src/main/kotlin/konan/worker/Worker.kt]: Done

**Pull request merged**